specs/merge/beacon-chain.md

```diff
@@ -0,0 +1,190 @@
+    - [`Transaction`](#transaction)
+    - [`ApplicationPayload`](#applicationpayload)
+- [Helper functions](#helper-functions)
+  - [Misc](#misc)
+    - [`is_transition_completed`](#is_transition_completed)
+    - [`is_transition_block`](#is_transition_block)
+  - [Block processing](#block-processing)
+    - [Application payload processing](#application-payload-processing)
+      - [`get_application_state`](#get_application_state)
+      - [`application_state_transition`](#application_state_transition)
+      - [`process_application_payload`](#process_application_payload)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This is a patch implementing executable beacon chain proposal. 
+It enshrines application execution and validity as a first class citizen at the core of the beacon chain.
+
+## Constants
+
+### Transition
+| Name | Value |
+| - | - |
+| `TRANSITION_TOTAL_DIFFICULTY` | _TBD_ |
+
+### Execution
+
+| Name | Value |
+| - | - |
+| `MAX_BYTES_PER_TRANSACTION_PAYLOAD` | `2**20` |
+| `MAX_APPLICATION_TRANSACTIONS` | `2**14` |
+| `BYTES_PER_LOGS_BLOOM` | `2**8` |
+
+
+## Containers
+
+### Extended containers
+
+*Note*: Extended SSZ containers inherit all fields from the parent in the original
+order and append any additional fields to the end.
+
+#### `BeaconBlockBody`
+
+*Note*: `BeaconBlockBody` fields remain unchanged other than the addition of `application_payload`.
+
+```python
+class BeaconBlockBody(phase0.BeaconBlockBody):
+    application_payload: ApplicationPayload  # [Added in Merge] application payload
+```
+
+#### `BeaconState`
+
+*Note*: `BeaconState` fields remain unchanged other than addition of `application_state_root` and `application_block_hash`. 
+
+
+```python
+class BeaconState(phase0.BeaconState):
+    # [Added in Merge] hash of the root of application state
+    application_state_root: Bytes32
+    # [Added in Merge] hash of recent application block
+    application_block_hash: Bytes32
+```
+
+### New containers
+
+#### `Transaction`
+
+Application transaction fields structured as an SSZ object for inclusion in an `ApplicationPayload` contained within a `BeaconBlock`.
+
+```python
+class Transaction(Container):
+    nonce: uint64
+    gas_price: uint256
+    gas_limit: uint64
+    recipient: Bytes20
+    value: uint256
+    input: List[Bytes1, MAX_BYTES_PER_TRANSACTION_PAYLOAD]
+    v: uint256
+    r: uint256
+    s: uint256
+```
+
+#### `ApplicationPayload`
+
+The application payload included in a `BeaconBlock`.
+
+```python
+class ApplicationPayload(Container):
+    block_hash: Bytes32  # Hash of application block
+    coinbase: Bytes20
+    state_root: Bytes32
+    gas_limit: uint64
+    gas_used: uint64
+    receipt_root: Bytes32
+    logs_bloom: Vector[Bytes1, BYTES_PER_LOGS_BLOOM]
+    difficulty: uint64  # Temporary field, will be removed later on
+    transactions: List[Transaction, MAX_APPLICATION_TRANSACTIONS]
```

Comment:
The only two fields in this class referenced by the spec are `block_hash` and `state_root`. Perhaps it would make things simpler to define `ApplicationPayload` as an abstract class?

Reply:
Application layer will need to be able to assemble an application block from an instance of `ApplicationPayload` and `parent_hash`. Thus, all these fields are important to be held by the beacon block body. Other application block fields (ethash, extra data) will be required to have zero values which will be verified by matching the `application_payload.block_hash` against the hash of assembled block on the application layer.